src/GNDStk/gnds-tree.hpp

```diff
@@ -0,0 +1,644 @@
+
+// -----------------------------------------------------------------------------
+// Tree (templated)
+// tree = Tree<>
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class MCON, // metadata container
+   template<class...> class CCON  // children container
+>
+class Tree {
+   static std::string static_str1;
+   static std::string static_str2;
+
+public:
+
+   // ------------------------
+   // Data
+   // ------------------------
+
+   // root
+   std::shared_ptr<Node<MCON,CCON>> root;
+
+
+   // ------------------------
+   // Miscellaneous functions
+   // ------------------------
+
+   // clear
+   void clear()
+   {
+      // smart pointer, so the rest of the tree is deleted as well
+      root = nullptr;
+   }
+
+   // empty
+   bool empty() const
+   {
+      return root == nullptr;
+   }
+
+   // start
+   Node<MCON,CCON> &start(
+      const std::string &type_str,
+      const std::string &gnds_str,
+      const std::string &param1 = static_str1,
+      const std::string &param2 = static_str2
+   ) {
+      // begin a new tree
+      clear();
+
+      // TYPE NODE: "xml", etc.
+      root = std::make_shared<Node<MCON,CCON>>();
+      root->name = type_str;
+      if (type_str == "xml") {
+         // xml
+         if (&param1 == &static_str1)
+            root->push("version","1.0");
+         else if (param1 != "")
+            root->push("version",param1);
+         if (&param2 == &static_str2)
+            root->push("encoding","UTF-8");
+         else if (param2 != "")
+            root->push("encoding",param2);
+      } else if (type_str == "hdf5") {
+         // fixme Write hdf5 case
+      } else {
+         // fixme Write json case
+      }
+
+      // PRIMARY GNDS NODE: "reactionSuite", etc.
+      Node<MCON,CCON> &gnds_node = root->push();
+      gnds_node.name = gnds_str;
+      return gnds_node;
+   }
+
+   // normalize
+   void normalize()
```

Comment:
It relates to my (not-entirely-implemented) ideas about taking "faked attributes" in JSON (because JSON doesn't have real attributes) and transforming them into proper attributes for the internal tree structure. So, in some sense, "normalizing" a JSON file from its wild, untamed self. That code needs some work anyway, and I have no problem changing the names of relevant functions if the present names bother anyone.

Reply:
Perhaps I'll have, say, attributes2children and children2attributes to deal with file formats that don't support attributes. More work needs to be done on all the JSON stuff to see how this all pans out.